{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import treeswift\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readFASTA(path):\n",
    "    '''\n",
    "    read FASTA stream and return (ID,seq) dictionary\n",
    "    '''\n",
    "    stream = open(path)\n",
    "    seqs = {}\n",
    "    name = None\n",
    "    seq = ''\n",
    "    for line in stream:\n",
    "        l = line.strip()\n",
    "        if len(l) == 0:\n",
    "            continue\n",
    "        if l[0] == '>':\n",
    "            if name is not None:\n",
    "                assert len(seq) != 0, \"Malformed FASTA\"\n",
    "                seqs[name] = seq.upper()\n",
    "            name = l[1:]\n",
    "            assert name not in seqs, \"Duplicate sequence ID: %s\" % name\n",
    "            seq = ''\n",
    "        else:\n",
    "            seq += l\n",
    "    assert name is not None and len(seq) != 0, \"Malformed FASTA\"\n",
    "    seqs[name] = seq.upper()\n",
    "    return seqs\n",
    "\n",
    "\n",
    "# get lineage of Hu-1 aligned sequence\n",
    "def get_lineage(seq):\n",
    "    '''\n",
    "    Get the lineage of the sequence based on the positions of the mutations at 8782 and 28144\n",
    "    '''\n",
    "    if seq[8782-1] == 'C' and seq[28144-1] == 'T':\n",
    "        lineage = 'B'\n",
    "    elif seq[8782-1] == 'T' and seq[28144-1] == 'C':\n",
    "        lineage = 'A'\n",
    "    elif seq[8782-1] == 'C' and seq[28144-1] == 'C':\n",
    "        lineage = 'CC'\n",
    "    elif seq[8782-1] == 'T' and seq[28144-1] == 'T':\n",
    "        lineage = 'TT'\n",
    "    else:\n",
    "        lineage = 'unknown'\n",
    "    return lineage\n",
    "\n",
    "\n",
    "# compare 2 sequences and return the mutations (e.g., C8782T)\n",
    "def compare_seqs(ref, seq):\n",
    "    mut_indices = []\n",
    "    for index, nt in enumerate(seq):\n",
    "        if (ref[index] != seq[index]) and ref[index] in 'ACGT' and seq[index] in 'ACGT':\n",
    "            mut_indices.append(ref[index] + str(index + 1) + seq[index])\n",
    "    return mut_indices\n",
    "\n",
    "\n",
    "def read_trees(path, recCA=False, exclude_taxon='ghost'):\n",
    "    '''\n",
    "    Read in the trees file and return a dictionary with the state as the key and the tree as the value\n",
    "    '''\n",
    "    trees = treeswift.read_tree_nexus(path)\n",
    "    # create an updated trees dictionary after removing \"STATE_\" from each key \n",
    "    updated_trees = {}\n",
    "    for key in trees:\n",
    "        if 'STATE_' in key:\n",
    "            tree = trees[key]\n",
    "            if recCA == False:\n",
    "                updated_trees[int(key[6:])] = tree\n",
    "            elif recCA == True: # if recCA, extract subtree that excludes ghost lineage\n",
    "                new_tree = tree.extract_tree_without([exclude_taxon])\n",
    "                updated_trees[int(key[6:])] = new_tree\n",
    "        else:\n",
    "            updated_trees[key] = trees[key]\n",
    "    return updated_trees\n",
    "\n",
    "\n",
    "def read_states(path):\n",
    "    '''\n",
    "    Read in the states file and return a dataframe with the state and the sequence\n",
    "    '''\n",
    "    states = pd.read_csv(path, sep='\\t', header=None, comment='#')\n",
    "    states.columns = ['state', 'sequence']\n",
    "    try:\n",
    "        states = states[~states['state'].str.contains('state')]\n",
    "    except:\n",
    "        # states = states\n",
    "        pass \n",
    "    states['state'] = states['state'].apply(lambda x: int(x))\n",
    "    states['lineage'] = states['sequence'].apply(lambda x: get_lineage(x))\n",
    "    return states\n",
    "\n",
    "\n",
    "def get_lineage_mrca(tree, lineage_dict):\n",
    "    '''\n",
    "    Returns a dictionary with the lineage as the key and the MRCA of the taxa with that lineage as the value\n",
    "    \n",
    "    Parameters:\n",
    "    tree (Tree): the tree\n",
    "    lineage_dict (Dict): a dictionary with the lineage as the key and all taxa with that lineage as the value\n",
    "    \n",
    "    Returns:\n",
    "    Dict: a dictionary with the lineage as the key and the MRCA of the taxa with that lineage as the value\n",
    "    '''\n",
    "    mrca_dict = {}\n",
    "    for key in lineage_dict:\n",
    "        if key == 'unknown':\n",
    "            continue\n",
    "        taxa = lineage_dict[key]\n",
    "        mrca = tree.mrca(taxa)\n",
    "        mrca_dict[key] = mrca\n",
    "\n",
    "        if key == 'TT':\n",
    "            for taxon in taxa:\n",
    "                mrca_dict[taxon] = tree.mrca([taxon]) # this is the node itself; it's for compatibility because there isn't a function in treeswift that returns the node given its label\n",
    "    return mrca_dict\n",
    "\n",
    "\n",
    "def subtree_taxa(tree, node, taxa):\n",
    "    '''\n",
    "    Returns whether the taxa are in the subtree with the given node as the root\n",
    "\n",
    "    Parameters:\n",
    "    tree (Tree): the tree\n",
    "    node (Node): the root of the subtree\n",
    "    taxa (List): the taxa to check\n",
    "\n",
    "    Returns:\n",
    "    Boolean: whether the taxa are in the subtree\n",
    "    '''\n",
    "\n",
    "    # subtree = tree.extract_subtree(node)\n",
    "    subtree_taxa = {n.label for n in node.traverse_leaves()} # this is traversing the leaves below this node\n",
    "    return all([taxon in subtree_taxa for taxon in taxa])\n",
    "\n",
    "\n",
    "def locate_tt(tree, mrca_lineage, lineage_dict, state=None):\n",
    "    '''\n",
    "    Determines whether the T/T lineage is between A and B, or if it is a descendant of A or B\n",
    "    \n",
    "    Parameters:\n",
    "    tree (Tree): the tree\n",
    "    mrca_lineage (String): the lineage of the MRCA of A and B\n",
    "    lineage_dict (Dict): a dictionary with the lineage as the key and all taxa with that lineage as the value\n",
    "    state (String): the state of the tree\n",
    "    \n",
    "    Returns:\n",
    "    Boolean: whether A is monophyletic\n",
    "    Boolean: whether B is monophyletic\n",
    "    Boolean: whether T/T is monophyletic\n",
    "    Boolean: when T/T is a true intermediate, the parent haplotype is 'intermediate'; otherwise, the parent haplotypes of T/T\n",
    "    '''\n",
    "\n",
    "    lineage_mrca_dict = get_lineage_mrca(tree, lineage_dict)\n",
    "    # a_size, b_size = tree.extract_subtree(lineage_mrca_dict['A']).num_nodes(internal=False), tree.extract_subtree(lineage_mrca_dict['B']).num_nodes(internal=False)\n",
    "    a_size, b_size = len(list(lineage_mrca_dict['A'].traverse_leaves())), len(list(lineage_mrca_dict['B'].traverse_leaves()))\n",
    "\n",
    "    num_leaves = len(list(tree.traverse_leaves())) # quick way to check whether A or B is not monophyletic but also not the root\n",
    "    if a_size > 285 and a_size < num_leaves:\n",
    "        print('A is not monophyletic, not the root, and evolved multiple times!', state)\n",
    "    if b_size > 585 and b_size < num_leaves:\n",
    "        print('B is not monophyletic, not the root, and evolved multiple times!', state)\n",
    "        \n",
    "    tt_monophyletic = True\n",
    "    true_intermediate = False\n",
    "    \n",
    "    # determine if T/T is monophyletic and whether to examine parents of individual T/T genomes or parent of the MRCA of T/T genomes\n",
    "    tt_genomes = lineage_dict['TT']\n",
    "    tt_options = ['TT']\n",
    "    # if tree.extract_subtree(lineage_mrca_dict['TT']).num_nodes(internal=False) != 2:\n",
    "    #     tt_monophyletic = False\n",
    "    #     tt_options = tt_genomes\n",
    "    if len(list(lineage_mrca_dict['TT'].traverse_leaves())) != 2:\n",
    "        tt_monophyletic = False\n",
    "        tt_options = tt_genomes\n",
    "\n",
    "    # determining whether T/T is between A and B; this does not depend on the monophyly of T/T -- only the parent of either a T/T genome or the parent of the MRCA of the T/T genomes\n",
    "    if mrca_lineage == 'A':\n",
    "        for tt_option in tt_options:\n",
    "            if lineage_mrca_dict['B'] in lineage_mrca_dict[tt_option].parent.children:\n",
    "                # print('A is ancestral haplotype, and B is a descendant of the parent of the mrca of T/T!', state)\n",
    "                true_intermediate = True\n",
    "                break\n",
    "\n",
    "    elif mrca_lineage == 'B':\n",
    "        for tt_option in tt_options:\n",
    "            if lineage_mrca_dict['A'] in lineage_mrca_dict[tt_option].parent.children:\n",
    "                # print('B is ancestral haplotype, and A is a descendant of the parent of the mrca of T/T!', state)\n",
    "                true_intermediate = True\n",
    "                break\n",
    "\n",
    "    elif mrca_lineage == 'TT':\n",
    "        for tt_option in tt_options:\n",
    "            if lineage_mrca_dict['A'] in lineage_mrca_dict[tt_option].parent.children:\n",
    "                # print('T/T is ancestral haplotype, and A is a descendant of the parent of the mrca of T/T!', state)\n",
    "                true_intermediate = True\n",
    "                break\n",
    "            elif lineage_mrca_dict['B'] in lineage_mrca_dict[tt_option].parent.children:\n",
    "                # print('T/T is ancestral haplotype, and B is a descendant of the parent of the mrca of T/T!', state)\n",
    "                true_intermediate = True\n",
    "                break\n",
    "    \n",
    "    elif mrca_lineage == 'CC':\n",
    "        for tt_option in tt_options:\n",
    "            if lineage_mrca_dict['A'] in lineage_mrca_dict[tt_option].parent.children:\n",
    "                # print('C/C is ancestral haplotype, and A is a descendant of the parent of the mrca of T/T!', state)\n",
    "                true_intermediate = True\n",
    "                break\n",
    "            elif lineage_mrca_dict['B'] in lineage_mrca_dict[tt_option].parent.children:\n",
    "                # print('C/C is ancestral haplotype, and B is a descendant of the parent of the mrca of T/T!', state)\n",
    "                true_intermediate = True\n",
    "                break\n",
    "\n",
    "    # if T/T is a true intermediate, then the parent haplotypes are 'intermediate'\n",
    "    # otherwise, determine the parent haplotype(s) of T/T -- there are two T/T genomes, so there could be up to two parent haplotypes\n",
    "    if true_intermediate:\n",
    "        parent_haplotypes = 'intermediate'\n",
    "    else:\n",
    "        parent_haplotypes = []\n",
    "        if tt_monophyletic:\n",
    "            part_of_a = subtree_taxa(tree, lineage_mrca_dict['A'], tt_genomes)\n",
    "            part_of_b = subtree_taxa(tree, lineage_mrca_dict['B'], tt_genomes)\n",
    "            if part_of_a and not part_of_b:\n",
    "                parent_haplotypes.append('A')\n",
    "            elif part_of_b and not part_of_a:\n",
    "                parent_haplotypes.append('B')\n",
    "            elif part_of_a and part_of_b: # this is the case where there is an A or B polytomy at the base of the tree; the smaller clade is the parent haplotype\n",
    "                if a_size > b_size: # check with Marc\n",
    "                    parent_haplotypes.append('B')\n",
    "                elif b_size > a_size:\n",
    "                    parent_haplotypes.append('A')\n",
    "                else:\n",
    "                    print('Need to check this case!', state) # something is wrong if the sizes are equal\n",
    "            else:\n",
    "                if mrca_lineage == 'A': # ancestral haplotype of A and then these occur on the branch leading to the MRCA of A\n",
    "                    parent_haplotypes.append('A')\n",
    "                elif mrca_lineage == 'B': # ancestral haplotype of B and then these occur on the branch leading to the MRCA of B\n",
    "                    parent_haplotypes.append('B')\n",
    "                elif mrca_lineage == 'TT': # should never reach this case\n",
    "                    import ipdb; ipdb.set_trace()\n",
    "                    parent_haplotypes.append('TT')\n",
    "                elif mrca_lineage == 'CC': # should never reach this case\n",
    "                    import ipdb; ipdb.set_trace()\n",
    "                    parent_haplotypes.append('CC')\n",
    "            if len(parent_haplotypes) == 0:\n",
    "                parent_haplotypes = ['unknown']\n",
    "        else:\n",
    "            for tt in tt_genomes:\n",
    "                part_of_a = subtree_taxa(tree, lineage_mrca_dict['A'], [tt])\n",
    "                part_of_b = subtree_taxa(tree, lineage_mrca_dict['B'], [tt])\n",
    "                if part_of_a and not part_of_b:\n",
    "                    parent_haplotypes.append('A')\n",
    "                elif part_of_b and not part_of_a:\n",
    "                    parent_haplotypes.append('B')\n",
    "                elif part_of_a and part_of_b: # this is the case where there is an A or B polytomy at the base of the tree; the smaller clade is the parent haplotype\n",
    "                    if a_size > b_size:\n",
    "                        if a_size != tree.num_nodes(internal=False):\n",
    "                            print('Need to check this case, a_size does not equal tree num nodes!', state) # something is wrong if size of A is not equal to the size of the tree\n",
    "                            # import ipdb; ipdb.set_trace()\n",
    "                            tree.write_tree_nexus('%s_tree.nexus' % state)\n",
    "                        parent_haplotypes.append('B')\n",
    "                    elif b_size > a_size:\n",
    "                        if b_size != tree.num_nodes(internal=False):\n",
    "                            print('Need to check this case, b_size does not equal tree num nodes!', state) # something is wrong if size of B is not equal to the size of the tree\n",
    "                            # import ipdb; ipdb.set_trace()\n",
    "                            tree.write_tree_nexus('%s_tree.nexus' % state)\n",
    "                        parent_haplotypes.append('A')\n",
    "                    else:\n",
    "                        print('Need to check this case!', state) # something is wrong if the sizes are equal\n",
    "                else:\n",
    "                    # import ipdb; ipdb.set_trace()\n",
    "                    if mrca_lineage == 'A': # ancestral haplotype of A and then these occur on the branch leading to the MRCA of A \n",
    "                        parent_haplotypes.append('A')\n",
    "                    elif mrca_lineage == 'B': # ancestral haplotype of B and then these occur on the branch leading to the MRCA of B\n",
    "                        parent_haplotypes.append('B')\n",
    "                    elif mrca_lineage == 'TT': # should never reach this case\n",
    "                        import ipdb; ipdb.set_trace()\n",
    "                        parent_haplotypes.append('TT')\n",
    "                    elif mrca_lineage == 'CC': # should never reach this case\n",
    "                        import ipdb; ipdb.set_trace()\n",
    "                        parent_haplotypes.append('CC')\n",
    "            if len(parent_haplotypes) != 2:\n",
    "                # import ipdb; ipdb.set_trace()\n",
    "                parent_haplotypes = ['unknown']\n",
    "        parent_haplotypes = ','.join(sorted(set(parent_haplotypes)))\n",
    "\n",
    "    return tt_monophyletic, parent_haplotypes\n",
    "\n",
    "\n",
    "def states_tt_results(trees, lineage_dict, states_path):\n",
    "    '''\n",
    "    Run the analysis on the trees and states\n",
    "    \n",
    "    Parameters:\n",
    "    trees (Dict): a dictionary with the state as the key and the tree as the value\n",
    "    lineage_dict (Dict): a dictionary with the lineage as the key and all taxa with that lineage as the value\n",
    "    states_path (String): the path to the states file\n",
    "    \n",
    "    Returns:\n",
    "    DataFrame: the states dataframe with the tt_between_ab column\n",
    "    '''\n",
    "    states = read_states(states_path)\n",
    "\n",
    "    tt_monophyletic_results = []\n",
    "    tt_ancestry_results = []\n",
    "    fails = []\n",
    "    for state in trees:\n",
    "        if type(state) != int:\n",
    "            continue\n",
    "\n",
    "        mrca_lineage = states[states['state'] == state]['lineage'].values[0]\n",
    "        # mrca_lineage = states[states['state'] == state]['lineage'].item()\n",
    "        try:\n",
    "            tt_monophyletic, tt_ancestry = locate_tt(trees[state], mrca_lineage, lineage_dict, state)\n",
    "            tt_monophyletic_results.append(tt_monophyletic)\n",
    "            tt_ancestry_results.append(tt_ancestry)\n",
    "            if tt_ancestry == 'unknown':\n",
    "                print('Unknown, need to check this case!', state, mrca_lineage, tt_monophyletic, tt_ancestry)\n",
    "                # import ipdb; ipdb.set_trace()\n",
    "                fails.append([state, trees[state], mrca_lineage, tt_monophyletic, tt_ancestry])\n",
    "        except:\n",
    "            print('Error in state', state, 'skipping')\n",
    "            # import ipdb; ipdb.set_trace()\n",
    "            tt_monophyletic_results.append('ERROR')\n",
    "            tt_ancestry.append('ERROR')\n",
    "            fails.append([state, trees[state], mrca_lineage, 'ERROR', 'ERROR', 'ERROR'])\n",
    "    \n",
    "    states['tt_monophyletic'] = tt_monophyletic_results\n",
    "    states['tt_ancestry'] = tt_ancestry_results\n",
    "    \n",
    "    return states, pd.DataFrame(fails, columns=['state', 'tree', 'mrca_lineage', 'tt_monophyletic', 'tt_ancestry'])\n",
    "\n",
    "\n",
    "def run_analysis(trees_path, lineage_dict, states_path, recCA=False, exclude_taxon='ghost', burnin=0.1):\n",
    "    '''\n",
    "    Run the analysis on the trees and states\n",
    "    \n",
    "    Parameters:\n",
    "    trees_path (String): the path to the trees file\n",
    "    lineage_dict (Dict): a dictionary with the lineage as the key and all taxa with that lineage as the value\n",
    "    states_path (String): the path to the states file\n",
    "    recCA (Boolean): whether the trees are recCA trees\n",
    "    exclude_taxon (String): the taxon to exclude from the trees\n",
    "    \n",
    "    Returns:\n",
    "    DataFrame: the states dataframe with the tt_between_ab column\n",
    "    '''\n",
    "    trees = read_trees(trees_path, recCA=recCA, exclude_taxon=exclude_taxon)\n",
    "    states, fails = states_tt_results(trees, lineage_dict, states_path)\n",
    "    states = states[states['state'] > (burnin * states['state'].max())]\n",
    "    states.columns = ['state', 'sequence', 'ancestral_haplotype', 'tt_monophyletic', 'tt_ancestry']\n",
    "\n",
    "    return states[['state', 'ancestral_haplotype', 'tt_monophyletic', 'tt_ancestry']], fails\n",
    "\n",
    "\n",
    "def run_all_analyses(trees_paths, states_paths, fasta_path, recCA=False, exclude_taxon='ghost', burnin=0.1, redo=False):\n",
    "    '''\n",
    "    Run the analysis on the trees and states\n",
    "    \n",
    "    Parameters:\n",
    "    trees_paths (List): a list of paths to the trees files\n",
    "    states_path (String): the path to the states file\n",
    "    recCA (Boolean): whether the trees are recCA trees\n",
    "    exclude_taxon (String): the taxon to exclude from the trees\n",
    "    '''\n",
    "    fasta = readFASTA(fasta_path)\n",
    "\n",
    "    lineage_dict = {}\n",
    "    for key in fasta:\n",
    "        lin = get_lineage(fasta[key])\n",
    "        if lin in lineage_dict:\n",
    "            lineage_dict[lin].append(key)\n",
    "        else:\n",
    "            lineage_dict[lin] = [key]\n",
    "    \n",
    "    for index, trees_path in enumerate(trees_paths):\n",
    "        if index != 2:\n",
    "            continue\n",
    "        intermediate_results_path = trees_path.replace('.trees', '.intermediatesResults.csv')\n",
    "        intermediate_fails_path = trees_path.replace('.trees', '.intermediatesFails.csv')\n",
    "        if os.path.exists(intermediate_results_path) and not redo:\n",
    "            print('Intermediate results already exist for', trees_path, 'skipping')\n",
    "            continue\n",
    "\n",
    "        print('Running analysis on', trees_path)\n",
    "        states_path = states_paths[index]\n",
    "        states, fails = run_analysis(trees_path, lineage_dict, states_path, recCA=recCA, exclude_taxon=exclude_taxon, burnin=burnin)\n",
    "        states.to_csv(intermediate_results_path, index=False)\n",
    "        fails.to_csv(intermediate_fails_path, index=False)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run analysis (subset/resampled and full analyses)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_path = '../data/2024.WHO_Lv2024.masked.noAmbigs.fasta'\n",
    "\n",
    "unconstrained_trees_paths = ['../market_unconstrained/run1/global.Lv2024.masked.GLM.HKY.trees',\n",
    "                            '../market_unconstrained/run2/global.Lv2024.masked.GLM.HKY.trees',\n",
    "                            '../market_unconstrained/run3/global.Lv2024.masked.GLM.HKY.trees',\n",
    "                            '../market_unconstrained/run4/global.Lv2024.masked.GLM.HKY.trees']\n",
    "\n",
    "\n",
    "unconstrained_states_paths = ['../market_unconstrained/run1/global.Lv2024.masked.GLM.HKY.asr.states.log',\n",
    "                                '../market_unconstrained/run2/global.Lv2024.masked.GLM.HKY.asr.states.log',\n",
    "                                '../market_unconstrained/run3/global.Lv2024.masked.GLM.HKY.asr.states.log',\n",
    "                                '../market_unconstrained/run4/global.Lv2024.masked.GLM.HKY.asr.states.log']\n",
    "\n",
    "recCA_trees_paths = ['../market_recCA/run1/global.Lv2024.masked.GLM.HKY.recCA_GARD.trees',\n",
    "                    '../market_recCA/run2/global.Lv2024.masked.GLM.HKY.recCA_GARD.trees',\n",
    "                    '../market_recCA/run3/global.Lv2024.masked.GLM.HKY.recCA_GARD.trees',\n",
    "                    '../market_recCA/run4/global.Lv2024.masked.GLM.HKY.recCA_GARD.trees']\n",
    "\n",
    "recCA_states_paths = ['../market_recCA/run1/global.Lv2024.masked.GLM.HKY.recCA_GARD.states.log',\n",
    "                        '../market_recCA/run2/global.Lv2024.masked.GLM.HKY.recCA_GARD.states.log',\n",
    "                        '../market_recCA/run3/global.Lv2024.masked.GLM.HKY.recCA_GARD.states.log',\n",
    "                        '../market_recCA/run4/global.Lv2024.masked.GLM.HKY.recCA_GARD.states.log']\n",
    "\n",
    "run_all_analyses(unconstrained_trees_paths, unconstrained_states_paths, fasta_path, recCA=False, exclude_taxon='ghost', burnin=0.1, redo=True)\n",
    "run_all_analyses(recCA_trees_paths, recCA_states_paths, fasta_path, recCA=True, exclude_taxon='ghost', burnin=0.1, redo=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load and examine results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_results(results_paths):\n",
    "    results = pd.read_csv(results_paths[0])\n",
    "    for path in results_paths[1:]:\n",
    "        results = pd.concat([results, pd.read_csv(path)])\n",
    "    return results\n",
    "\n",
    "\n",
    "def convert_table(df):\n",
    "    '''\n",
    "    Summarize the results of the analysis, broken down by ancestral haplotype and the parent haplotype of the T/T genomes\n",
    "    '''\n",
    "    \n",
    "    # for each unique value of tt_child_of, make a new column and give it a 1 or 0 based on the value of tt_child_of for that row\n",
    "    df_edited = pd.get_dummies(df, columns=['tt_ancestry'])\n",
    "    # get rid of state column\n",
    "    df_edited = df_edited.drop(columns=['state'])\n",
    "    # group the results by ancestral haplotype and divide by frequency of that haplotype using agg\n",
    "    df_grouped = df_edited.groupby('ancestral_haplotype').agg('sum')\n",
    "    # add column for total number of trees for each ancestral haplotype\n",
    "    df_grouped['total'] = df_edited.groupby('ancestral_haplotype').size()\n",
    "    # add row of sum of each column\n",
    "    df_grouped.loc['total'] = df_grouped.sum()\n",
    "    # reorder to make total column first \n",
    "    cols = df_grouped.columns.tolist()\n",
    "    cols = cols[-1:] + cols[:-1]\n",
    "    df_grouped = df_grouped[cols]\n",
    "\n",
    "    # rename the columns\n",
    "    new_col_names = []\n",
    "    for col in df_grouped.columns.tolist():\n",
    "        col = col.split('ancestry')[-1].split('_')[-1]\n",
    "        if 'total' in col or 'monophyletic' in col or 'intermediate' in col:\n",
    "            col = col.capitalize()\n",
    "        else:\n",
    "            col = 'Derived from ' + ' and '.join(col.split(','))\n",
    "        new_col_names.append(col)\n",
    "    reordered_col_names = [x for x in ['Total', 'Monophyletic', 'Intermediate', 'Derived from A', 'Derived from B', 'Derived from A and B'] if x in new_col_names]\n",
    "    \n",
    "    df_grouped.columns = new_col_names\n",
    "    df_grouped = df_grouped[reordered_col_names]\n",
    "    \n",
    "    # rename the indices\n",
    "    df_grouped.index.name = 'Ancestral haplotype'\n",
    "    df_grouped = df_grouped.rename(index={'A': 'A', 'B': 'B', 'TT': 'T/T', 'CC': 'C/C', 'total': 'Total'})\n",
    "\n",
    "    # double check that all columns except for \"Monophyletic\" add up to the total column\n",
    "    for index in df_grouped.index:\n",
    "        total = df_grouped.loc[index, 'Total']\n",
    "        sum_columns = df_grouped.loc[index, 'Intermediate':].sum()\n",
    "        assert total == sum_columns, 'Error in ' + index\n",
    "\n",
    "    # for every column besides the total column, divide by the total number of trees for each ancestral haplotype, return the count with the percentage in parenthesis, but only to 3 decimal points\n",
    "    for col in df_grouped.columns:\n",
    "        if col != 'Total':\n",
    "            df_grouped[col] = df_grouped[col].astype(str) + ' (' + (df_grouped[col]/df_grouped['Total']*100).round(1).astype(str) + '%)'\n",
    "\n",
    "    return df_grouped\n",
    "\n",
    "\n",
    "unconstrained_results_path = ['../market_unconstrained/run1/global.Lv2024.masked.GLM.HKY.intermediatesResults.csv',\n",
    "                              '../market_unconstrained/run2/global.Lv2024.masked.GLM.HKY.intermediatesResults.csv',\n",
    "                            '../market_unconstrained/run3/global.Lv2024.masked.GLM.HKY.intermediatesResults.csv',\n",
    "                            '../market_unconstrained/run4/global.Lv2024.masked.GLM.HKY.intermediatesResults.csv']\n",
    "\n",
    "recCA_results_path = ['../market_recCA/run1/global.Lv2024.masked.GLM.HKY.recCA_GARD.intermediatesResults.csv',\n",
    "                    '../market_recCA/run2/global.Lv2024.masked.GLM.HKY.recCA_GARD.intermediatesResults.csv',\n",
    "                    '../market_recCA/run3/global.Lv2024.masked.GLM.HKY.recCA_GARD.intermediatesResults.csv',\n",
    "                    '../market_recCA/run4/global.Lv2024.masked.GLM.HKY.recCA_GARD.intermediatesResults.csv']\n",
    "\n",
    "\n",
    "unconstrained_results = combine_results(unconstrained_results_path)\n",
    "recCA_results = combine_results(recCA_results_path)\n",
    "\n",
    "unconstrained_results_table = convert_table(unconstrained_results)\n",
    "recCA_results_table = convert_table(recCA_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Total</th>\n",
       "      <th>Monophyletic</th>\n",
       "      <th>Intermediate</th>\n",
       "      <th>Derived from A</th>\n",
       "      <th>Derived from B</th>\n",
       "      <th>Derived from A and B</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ancestral haplotype</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>737</td>\n",
       "      <td>734 (99.6%)</td>\n",
       "      <td>0 (0.0%)</td>\n",
       "      <td>260 (35.3%)</td>\n",
       "      <td>476 (64.6%)</td>\n",
       "      <td>1 (0.1%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>5390</td>\n",
       "      <td>5367 (99.6%)</td>\n",
       "      <td>20 (0.4%)</td>\n",
       "      <td>1817 (33.7%)</td>\n",
       "      <td>3546 (65.8%)</td>\n",
       "      <td>7 (0.1%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C/C</th>\n",
       "      <td>2655</td>\n",
       "      <td>2643 (99.5%)</td>\n",
       "      <td>0 (0.0%)</td>\n",
       "      <td>867 (32.7%)</td>\n",
       "      <td>1782 (67.1%)</td>\n",
       "      <td>6 (0.2%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T/T</th>\n",
       "      <td>218</td>\n",
       "      <td>216 (99.1%)</td>\n",
       "      <td>1 (0.5%)</td>\n",
       "      <td>76 (34.9%)</td>\n",
       "      <td>140 (64.2%)</td>\n",
       "      <td>1 (0.5%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>9000</td>\n",
       "      <td>8960 (99.6%)</td>\n",
       "      <td>21 (0.2%)</td>\n",
       "      <td>3020 (33.6%)</td>\n",
       "      <td>5944 (66.0%)</td>\n",
       "      <td>15 (0.2%)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Total  Monophyletic Intermediate Derived from A  \\\n",
       "Ancestral haplotype                                                    \n",
       "A                      737   734 (99.6%)     0 (0.0%)    260 (35.3%)   \n",
       "B                     5390  5367 (99.6%)    20 (0.4%)   1817 (33.7%)   \n",
       "C/C                   2655  2643 (99.5%)     0 (0.0%)    867 (32.7%)   \n",
       "T/T                    218   216 (99.1%)     1 (0.5%)     76 (34.9%)   \n",
       "Total                 9000  8960 (99.6%)    21 (0.2%)   3020 (33.6%)   \n",
       "\n",
       "                    Derived from B Derived from A and B  \n",
       "Ancestral haplotype                                      \n",
       "A                      476 (64.6%)             1 (0.1%)  \n",
       "B                     3546 (65.8%)             7 (0.1%)  \n",
       "C/C                   1782 (67.1%)             6 (0.2%)  \n",
       "T/T                    140 (64.2%)             1 (0.5%)  \n",
       "Total                 5944 (66.0%)            15 (0.2%)  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Total</th>\n",
       "      <th>Monophyletic</th>\n",
       "      <th>Intermediate</th>\n",
       "      <th>Derived from A</th>\n",
       "      <th>Derived from B</th>\n",
       "      <th>Derived from A and B</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ancestral haplotype</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>8216</td>\n",
       "      <td>8140 (99.1%)</td>\n",
       "      <td>8 (0.1%)</td>\n",
       "      <td>2834 (34.5%)</td>\n",
       "      <td>5348 (65.1%)</td>\n",
       "      <td>26 (0.3%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>83</td>\n",
       "      <td>83 (100.0%)</td>\n",
       "      <td>0 (0.0%)</td>\n",
       "      <td>26 (31.3%)</td>\n",
       "      <td>57 (68.7%)</td>\n",
       "      <td>0 (0.0%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C/C</th>\n",
       "      <td>539</td>\n",
       "      <td>535 (99.3%)</td>\n",
       "      <td>0 (0.0%)</td>\n",
       "      <td>191 (35.4%)</td>\n",
       "      <td>347 (64.4%)</td>\n",
       "      <td>1 (0.2%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T/T</th>\n",
       "      <td>162</td>\n",
       "      <td>160 (98.8%)</td>\n",
       "      <td>1 (0.6%)</td>\n",
       "      <td>59 (36.4%)</td>\n",
       "      <td>102 (63.0%)</td>\n",
       "      <td>0 (0.0%)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>9000</td>\n",
       "      <td>8918 (99.1%)</td>\n",
       "      <td>9 (0.1%)</td>\n",
       "      <td>3110 (34.6%)</td>\n",
       "      <td>5854 (65.0%)</td>\n",
       "      <td>27 (0.3%)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Total  Monophyletic Intermediate Derived from A  \\\n",
       "Ancestral haplotype                                                    \n",
       "A                     8216  8140 (99.1%)     8 (0.1%)   2834 (34.5%)   \n",
       "B                       83   83 (100.0%)     0 (0.0%)     26 (31.3%)   \n",
       "C/C                    539   535 (99.3%)     0 (0.0%)    191 (35.4%)   \n",
       "T/T                    162   160 (98.8%)     1 (0.6%)     59 (36.4%)   \n",
       "Total                 9000  8918 (99.1%)     9 (0.1%)   3110 (34.6%)   \n",
       "\n",
       "                    Derived from B Derived from A and B  \n",
       "Ancestral haplotype                                      \n",
       "A                     5348 (65.1%)            26 (0.3%)  \n",
       "B                       57 (68.7%)             0 (0.0%)  \n",
       "C/C                    347 (64.4%)             1 (0.2%)  \n",
       "T/T                    102 (63.0%)             0 (0.0%)  \n",
       "Total                 5854 (65.0%)            27 (0.3%)  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(unconstrained_results_table)\n",
    "display(recCA_results_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save tables\n",
    "# unconstrained_results_table.to_csv('../market_unconstrained/intermediatesResultsTable.csv')\n",
    "# recCA_results_table.to_csv('../market_recCA/intermediatesResultsTable.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
